### DISCUSSION TITLE
SAFE procedure Phase 4 prior knowledge clarification

### DISCUSSION THREAD

**Discussion by RaySit** [asreview/asreview#1775] [Q&A] [answered]
We are working on a systematic review using the SAFE procedure, and we seek clarification regarding Phase 4 (the "quality evaluation" step).

The understanding is that the process for feeding active learning-based models "prior knowledge" (e.g., using ASReview) differs at each step:
- In Phase 2, we export a file from our reference manager and import it into ASReview as a database. None of the records have been screened yet, so we randomly screen a subset to become the “prior knowledge” for the models to use to generate the first iteration of the reference "stack".
- In Phase 3, we export our Phase 2 stack from ASReview and then import it as a database used to train a new model. ASReview automatically uses the records labeled "relevant" and "irrelevant" in Phase 2 as prior knowledge to train the Phase 3 model.
- In Phase 4, your manuscript indicates we should import only the irrelevant records from previous phases (actually, the text uses the term "previous phases", whereas the "cheat sheet" indicates the irrelevant records should come from Phase 2). For our team, this has raised the following questions:

1. Importing only records marked irrelevant suggests that ASReview will not have any "relevant" records to work off of in generating prior knowledge. Your manuscript "cheat sheet" indicates that prior knowledge for Phase 4 should come from "The 10 highest- and lowest-ranked papers [labeled irrelevant] from phase 2". How do we make this work?
2. Expanding on (1): To our knowledge, ASReview always puts irrelevant records on the bottom of the reference "stack". Here is a screenshot of the results of our Phase 3 exported as an Excel file. The highlighted parts are the irrelevant records, which are labeled 0 in column R ("included") and at the bottom of the list. 
![image](https://github.com/asreview/asreview/assets/77477538/fec7b158-9b00-4a00-9d6d-aba78b8ee5e4)

Inside this irrelevant stack, the records are, in order from top to bottom:
1. records labeled irrelevant in Phase 2 and used as prior knowledge in Phase 3 - these have a "1" in column S ("asreview_prior")
2. records labeled irrelevant in Phase 3 (with the "oldest" irrelevant-labeled records from Phase 3 going at the top of this section, and the "newest" irrelevant-labeled records from Phase 3 going to the bottom) - these have a "0" in column S ("asreview_prior")

Therefore, the 10 "highest-ranked" irrelevant records from this database would be those labeled irrelevant in Phase 2 and used as prior knowledge in Phase 3, while the 10 "lowest-ranked" irrelevant records will be the 10 records most recently labeled as "irrelevant" in Phase 3. This doesn't seem correct to us given that the "ranking" seems to reflect when the article was labeled irrelevant, rather than actual relevancy. Is this understanding correct? And, if so, can you help us understand what we might be doing wrong here?

**Comment by Rensvandeschoot** [comment, marked as answer]
It's great that you want to use the SAFE procedure! 

The main point is that any machine-learning method requires training data. For the active learning pipeline implemented in ASReview, you need at least one relevant and one irrelevant record. We also know that more complex models need more training data, for example, [Teijema et al.](https://doi.org/10.3389/frma.2023.1178181). But if you start with a new review project, you might have a large amount of labels (which might be different if you update an existing review). So, the question is how to find the training data or prior knowledge, as we call it. 

One method is to use target or cornerstone papers you already know should exist, plus some random irrelevant papers. However, we want to use such target papers to determine our stopping rule. Therefore, we proposed starting with randomly labeling papers until you randomly hit the first relevant paper. But any method will suffice, as long as you keep in mind that you want to start with the active learning pipeline as quickly as possible (to save time). Because you do not have many labels at the start, complex models should be avoided, and we propose to use a more lightweight model. 

After some time screening many papers, you have created a much larger training dataset. In [this paper ]( https://doi.org/10.3390/app14093842), we explain how the training set is used and grows with every human label. At some point, you will have enough labels for the more complex models, which will also help you find the more difficult papers that are relevant based on their context. So, when switching to the more complex models, you need as much training data as possible.

Because we want to be sure you did not exclude any relevant paper due to, for example, screening fatigue, you want an extra quality check, see also our paper on the [Noisy Label Filter ](https://doi.org/10.1186/s13643-024-02472-w) procedure. In this step, you want to screen only the excluded papers, but again, you need training data. It does not matter too much what prior knowledge you want to use. You can use all relevant papers from the previous step, but some might not be relevant since you still need to read the full text. Also, you need irrelevant papers as prior knowledge and those are exactly the papers you want to double-check, right? So, we used the highest / lowest-ranked papers from the previous step because these are predicted as the highest / lowest likely relevant as predicted by the machine learning model. You can also choose any other training set, but this selection seemed reasonable to us. 

The rank order column in the Excel file is based on the predicted relevance scores of the last iteration of the model. So, you must select the top number of records for the relevant set and the last few for the irrelevant set. 

Hopefully, this explanation helps, but feel free to ask more questions!!!! as a community we have to develop a new way of working with AI-aided tools, so such discussions are extremely useful!!

**Reply by RaySit** [reply]
Thank you, @Rensvandeschoot, for your clarifications. Based on your reply, I have two further questions:

(1) We attempted the following steps in phase 4 according to your instructions. Could you please verify if we are on the right track?
    1. We exported the entire dataset after completing Phase 3, saving it as an Excel sheet.
    2. In the Excel sheet, we deleted the non-top-ten relevant records and the unlabeled records, leaving only the top ten relevant records and all the irrelevant records.
    3. We then created a new project, set it to Validation Mode, and imported the Excel sheet as the dataset.
    4. For prior knowledge, we selected Random → Show 10 random relevant records, and labeled all our of top 10 previously-labeled relevant records as relevant again. We then manually searched for the bottom 10 irrelevant records and labeled them as irrelevant.
    5. We warmed up the AI and started screening, leaving us with only irrelevant records to screen (minus the 10 irrelevant records used as prior knowledge in Step 4 above).

(2) In your 2024 Systematic Reviews paper, the stopping rule for Phase 1 is described as "to screen a minimum number of records **OR** up to the point where at least one relevant record is found." However, in another sentence, it reads, "we propose to start by **LABELING** a random set of records for the training data (e.g., **1% of the total number of records**)." Does this mean we should generate a random set consisting of 1% of the total dataset, then screen until we find one relevant record, and proceed to calculate the Fraction of Relevant Records in the training set (FRR_t)? Or generate a random set consisting of 1% of the total dataset and screen that subset in its entirety (checking to ensure we have at least one relevant record) before moving on to subsequent steps?

**Reply by Rensvandeschoot** [reply]
1) Yes, indeed! We are working very hard on version 2 of ASReview, which will make this process much easier, but for now, this works!

2) You can take 1% or any other percentage. The value is arbitrary, but it should reflect a small number of records that can be used for training the first iteration of the model. The minimal training set size is two: 1 relevant + 1 irrelevant. We were also struggling with this first part if you want to start with a larger training set. Therefore, more [recently](https://www.crd.york.ac.uk/prospero/display_record.php?RecordID=546427), we combined the first phase of SAFE with the training/calibration phase of a review. That is, most often, you always start with training the screeners: screening a set of papers together, followed by a second training phase where each screener independently screens a batch and the results are discussed together. You could use these labels for the first phase of SAFE. 
For the training set, we took a subset of the oldest papers, the newest papers and some random papers in between (using a new function in [ASReview-Datatools](https://github.com/asreview/asreview-datatools?tab=readme-ov-file#sample)).

**Reply by RaySit** [reply]
Got it, thank you for the clarifications! Looking forward to work using version 2 someday in the future!

**Reply by RaySit** [reply]
Hello @Rensvandeschoot, I tried to use the `sample` function in the ASReview-Datatools to generate the training set for Phase 1. 
```
C:\Users\18922\Downloads>asreview data sample MY_OUTPUT.xlsx output_dataset.xslx 50
2024-07-15 14:31:02.860900: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-07-15 14:31:03.421251: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
Namespace(subcommand='data', version=False)
usage: asreview data [-h] [-V] [subcommand]

Tools for data preprocessing for ASReview.

positional arguments:
  subcommand     The datatool to launch. Available commands:

                 ['describe', 'dedup', 'convert', 'compose', 'vstack', 'snowball']

options:
  -h, --help     show this help message and exit
  -V, --version  show program's version number and exit
```

It seems like `sample` does not belong to the datatools. What is wrong here? How can I use it?

For now I just click random in the prior knowledge part and keep doing that until I have enough labeled records (1%). This allows the sample to be completely random, but it loses the stratified feature `sample` provides.

**Reply by Rensvandeschoot** [reply]
Can you try again after upgrading to version 1.3? (just released)

**Reply by RaySit** [reply]
@Rensvandeschoot Yes it worked. Thanks for the update!

A tiny typo in the [ASReview-Datatools](https://github.com/asreview/asreview-datatools?tab=readme-ov-file#sample) page:

In the `Sample` section, the following line of code is provided:

```
asreview data sample input_dataset.xlsx output_dataset.xslx 50
```

It should be `output_dataset.xlsx` instead.

**Reply by Rensvandeschoot** [reply]
@RaySit thanks! could you maybe start a pull request with this change? this way, you will get the credits and contribute to an open-source project always reflects good on your resume ;-)

**Reply by laurens88** [reply]
Thank you @RaySit, I corrected the mistake in a PR: https://github.com/asreview/asreview-datatools/pull/47

**Reply by Annasmits** [reply]
At this moment I have exactly the same question as RaySit posted in the opening of this thread. Unfortunately, I cannot follow(understand) your reply @Rensvandeschoot . I am a non-expert and am therefore trying to follow the steps from the SAFE procedure paper very closely for my large scale qualitative scoping review. (I've also spoken to @FelixWdm about my project before). 
At the moment it is going very well. I've completed phase 2 and will try to finish phase 3 this week. But I do not know which steps to follow to use the irrelevant papers from phase 3 (that is how I understood it from the paper) for phase 4. I do not understand how to find the highest and lowest ranked papers. Also I do not see how I can start a new project with only irrelevant papers for prior knowledge. Could someone try and explain the steps to me in (even more) laymen language?

**Comment by Annasmits** [comment]
@RaySit would you mind sharing how you've proceeded for SAFE 4? I still have trouble folllowing @Rensvandeschoot his explanation.
Should I export the entire set after fase3? (I've switched models between safe 2 and 3 in ASreview 2.0). Then select the 10 highest and 10 lowest references in the excel. Upload those + all remaining irrelevant labeled (both by computer and by me) references into a new project, then manually search for 10 hijghest and lowest from previous phase for prior knowledge and then start screening? #SAFE2.0

**Reply by Rensvandeschoot** [reply]
I recommend a much simpler procedure for SAFE Phase 4 in **ASReview v2**. Here is a step-by-step outline:

**Step 1:**
In ASReview v2, go to **export** and choose to export:

* The **excluded records**
* The **relevant records**

**Step 2:**
Open the exported file (Excel or your reference manager) and:

* For **excluded records**: keep them all, but you can remove the label `0`  (in Excel) or `ASReview_irrelevant` (RIS file format) because you want to label these records in the next phase. 
* For **included records**: keep only the ones you are **fully confident** about (delete those you're unsure about — you only need a few solid relevant ones for Phase 4).

**Step 3:**
Start a **new ASReview project**:

* Import this new dataset (excluded ones without label + included records with a label).
* The system will automatically recognize the **relevant records** as prior knowledge.

**Step 4:**
Begin screening:

* You will first see **random records**.
* As soon as you label at least one **irrelevant** record, the model will start learning.
* The AI will then present records ranked by likelihood of being relevant — this allows you to focus on any potentially **missed relevant records** from earlier phases.

👉 No need to manually pick 10 highest/lowest ranked records anymore — ASReview v2 is optimized to learn quickly from your selected relevant records in Phase 4.


(and the next feature in a future release should also accomodate this step of the SAFE procedure, but this is work in progress)

**Reply by Annasmits** [reply]
Thank you so much Rens! I feel a lot more confident now. Will try right away!
